<img src="https://www.th-koeln.de/img/logo.svg" style="float:right;" width="200">

# 12th exercise: <font color="#C70039">First Reinforcement Learning Game (*Frozen Lake*) using OpenAI Gym</font>
* Course: AML
* Lecturer: <a href="https://www.gernotheisenberg.de/">Gernot Heisenberg</a>
* Author of notebook: <a href="https://www.gernotheisenberg.de/">Gernot Heisenberg</a>. This notebook is based on the great post and notebook from [Rodolfo Mendes](https://morioh.com/p/18a96b9091d3).
* Author of notebook: <a href="https://www.gernotheisenberg.de/">Rubaiya Kabir Pranti</a>
* Matriculation Number: <a href="https://www.gernotheisenberg.de/"> 11146364</a>
* Date:   16.01.2024

<img src="https://launchyourintelligentapphome.files.wordpress.com/2019/05/frozenlake_legended.png?w=531" style="float: center;" width="600">

---------------------------------
**GENERAL NOTE 1**: 
Please make sure you are reading the entire notebook, since it contains a lot of information on your tasks (e.g. regarding the set of certain paramaters or a specific computational trick), and the written mark downs as well as comments contain a lot of information on how things work together as a whole. 

**GENERAL NOTE 2**: 
* Please, when commenting source code, just use English language only. 
* When describing an observation please use English language, too.
* This applies to all exercises throughout this course.

---------------------------------

### <font color="ce33ff">DESCRIPTION</font>:

#### OpenAI Gym
In this exercise you will be using Python and OpenAI Gym to develop your reinforcement learning algorithm. The Gym library is a collection of environments that can be used freely with the reinforcement learning algorithms.

Gym has a ton of environments ranging from simple text based games to Atari games like Breakout and Space Invaders. The library is intuitive to use and simple to install. Just run **pip install gym** and you are good to go! The link to Gym's installation instructions, requirements, and documentation is included in the description. 

Further reading about OpenAI Gym is available under https://www.gymlibrary.dev/.
This notebook is based on this great post and notebook from [Rodolfo Mendes](https://morioh.com/p/18a96b9091d3).

#### Frozen Lake
This description of the game is copied directly from Gym's website. 

*Winter is coming. You and your friends were tossing around a frisbee at the park when you made a wild throw that left the frisbee out in the middle of the lake. The water is mostly frozen, but there are a few holes where the ice has melted. If you step into one of those holes, you'll fall into the freezing water and die (Game over). At this time, there's an international frisbee shortage, so it's absolutely imperative that you navigate across the lake and retrieve the disc. However, the ice is slippery, so you won't always move in the direction you intend. The surface is described using a grid like the following:*

* SFFF
* FHFH
* FFFH
* HFFG

This grid is your environment! S is your (the agent's) starting point and it's safe. F represents the frozen surface and is also safe. H represents a hole and if your agent steps in a hole in the middle of a frozen lake, the game is over because the agent dies. Finally, G represents the goal, which is the space on the grid where the frisbee is located.

The agent can navigate *left, right, up, down* and the episode ends when the agent reaches the goal or falls in a hole. It receives a reward of **1** if it reaches the goal and **0** otherwise.

Here is the summary:
<img src="./images/FrozenLake.States.Rewards.png" style="float: center;" width="800">

---------------------------------

### <font color="FFC300">TASKS</font>:
The tasks that you need to work on within this notebook are always indicated below as bullet points. 
If a task is more challenging and consists of several steps, this is indicated as well. 
Make sure you have worked down the task list and commented your doings. 
This should be done by using markdown.<br> 
<font color=red>Make sure you don't forget to specify your name and your matriculation number in the notebook.</font>

**YOUR TASKS in this exercise are as follows**:
1. import the notebook to Google Colab or use your local machine.
2. make sure you specified you name and your matriculation number in the header below my name and date. 
    * set the date too and remove mine.
3. read the entire notebook carefully 
    * add comments whereever you feel it necessary for better understanding
    * run the notebook for the first time. 
4. install gym into your env!
5. You will train an agent to play the *Frozen Lake* game using Q-learning and you will get a playback of how the agent does after being trained.
6. Again the task: Your agent has to navigate the grid by staying on the frozen surface without falling into any holes until it reaches the frisbee. If it reaches the frisbee, it wins with a reward of plus one. If it falls in a hole, it loses and receives no points for the entire episode.
7. Your tasks are highlighted in the notebook (see below)
---------------------------------

### Imports 
import all important libs including gym

In [1]:
!pip install gym

In [2]:
import numpy as np
import gym
import random
import time
from   IPython.display import clear_output

In [3]:
print(gym.__version__)

0.26.2


### Creating the Environment
For creating your environment, just call *gym.make()* and pass a string of the name of the environment you want to set up. 
All the environments with their corresponding names you can use here are available on Gym's website (see above).
With this *env* object, you are able to query for information about the environment, sample states and actions, retrieve rewards and have your agent navigate the frozen lake. That is all made available to you conveniently with Gym.

In [4]:
env = gym.make("FrozenLake-v1")

### Creating the Q-Table
Now, construct your Q-table, and initialize all the Q-values to zero for each state-action pair.
The number of rows in the table is equivalent to the size of the state space in the environment, and the number of columns is equivalent to the size of the action space (see above). You can get this information using *env.observation_space.n* and *env.action_space.n* as shown below in the code. Then, you can use this information to build the Q-table and initialize it with zeros.

In [5]:
action_space_size = env.action_space.n
state_space_size = env.observation_space.n

q_table = np.zeros((state_space_size, action_space_size))

In [6]:
action_space_size

4

In [7]:
state_space_size

16

In [8]:
print(q_table)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


### Initializing Q-Learning hyperparameters
Now, we're going to create and initialize all the parameters needed to implement the Q-learning algorithm.

First, with *num_episodes*, you define the total number of episodes you want the agent to play during training. Then, with *max_steps_per_episode*, you define a maximum number of steps that your agent is allowed to take within a single episode. So, if by the 100th step, the agent has not reached the frisbee or fallen through a hole, then the episode will terminate with the agent receiving zero points.

Next, you will set your *learning_rate* and your *discount_rate* as well, which was represented with the symbol (lambda) in the course slides (keyword: discounted return G_t).

Now, the last four parameters are all related to the exploration-exploitation dilemma with respect to the epsilon-greedy policy. You are initializing your *exploration_rate* to **1** and setting the *max_exploration_rate* to **1** and a *min_exploration_rate* to **0.01**. The *max* and *min* are just bounds to how large or small your exploration rate can be. Remember, the exploration rate was represented with the symbol (epsilon) when discussed in the course slides.

Lastly, you will set the *exploration_decay_rate* to **0.01** to determine the rate at which the *exploration_rate* will decay.

**YOUR <font color="FFC300">TASK</font> in this exercise is as follows** (point 7 from the task list above):

All of the above parameters can change!
Your task is to create a *testplan* and tune all parameters by yourself and observe how they influence and change the performance of the algorithm. 
Make notes! They will help you during the exam.

In [9]:
num_episodes = 10000
max_steps_per_episode = 100

learning_rate = 0.2 #alpha
discount_rate = 1 #gamma

exploration_rate = 1
max_exploration_rate = 1
min_exploration_rate = 0.01
exploration_decay_rate = 0.0005

Create a list to hold all of the rewards you will get from each episode. 
By means of this you can observe how your game score changes over time.

In [10]:
rewards_all_episodes = []

In the following code section, the entire Q-learning algorithm is implemented as discussed in detail in the AML course. 
When this code is executed, this is exactly where the training will take place. 
* The first for-loop contains everything that happens within a single episode. 
* The second nested loop contains everything that happens for a single time-step.

Read all the red comments, as they contain lots of important information on the implementation.

## Q-learning algorithm where the training will take place

In [11]:
# Q-learning algorithm

# loop: for a single episode
for episode in range(num_episodes):
    # initialize 'new episode' parameters
    state, info = env.reset()
    ''' The done variable just keeps track of whether or not your episode is finished.
    Initialize it to False when first starting the episode and you will see later where 
    it will get updated to notify you when the episode is over.'''
    done = False
    
    ''' Keep track of the rewards within the current episode as well.
    Hence, set rewards_current_episode = 0 since you start 
    with no rewards at the beginning of each episode.'''
    rewards_current_episode = 0

    # nested loop: for a single time-step
    for step in range(max_steps_per_episode): 
        # Exploration-exploitation trade-off
        '''For each time-step within an episode set your exploration_rate_threshold 
        to a random number between 0 and 1. This will be used to determine whether 
        your agent will explore or exploit the environment in this time-step.'''
        exploration_rate_threshold = random.uniform(0, 1)
        # If the exploration_rate_threshold is greater than our predefined exploration_rate==1, 
        #then agent chooses the action that it believes has the highest Q-value for the current state. This is known as exploitation.
        if exploration_rate_threshold > exploration_rate:
            action = np.argmax(q_table[state,:]) 
        # if exploration_rate_threshold is less than or equal to exploration_rate, 
        # the agent explores the environment by selecting an action randomly
        # and may not necessarily be the 'best action' according to the agent's current knowledge
        else:
            action = env.action_space.sample() #randomly chosen action

        # Take new action
        '''After action is chosen, take that action by calling step() on your env object and 
        pass your action to it. The function step() returns a tuple containing the new state, 
        the reward for the action you took, whether or not the action ended the episode and 
        diagnostic information regarding the environment (helpful for debugging).'''
        new_state, reward, done, truncated, info = env.step(action)

        # Update Q-table for Q(s,a)
        '''Compare this implementation with the equation in the course slides.'''
        q_table[state, action] = q_table[state, action] * (1 - learning_rate) + learning_rate * (reward + discount_rate * np.max(q_table[new_state, :]))
        
        '''Set your current state to the new_state that was returned when taking the last action 
        and then update the rewards from your current episode by adding the reward you received 
        for your previous action.'''
        # Set new state
        state = new_state
        # Add new reward 
        rewards_current_episode += reward 
        '''Then, check to see if your last action ended the episode 
        (game over by agent stepping in a hole or reaching the goal)! 
        If the action did end the episode, then jump out of this loop and start the next episode.
        Otherwise, transition to the next time-step.'''
        if done == True: 
            break
            
    # Exploration rate decay
    '''Once an episode is finished, you need to update your exploration_rate using exponential decay, 
    which just means that the exploration rate decays at a rate proportional to its current value. 
    You can decay the exploration_rate using the formula below, which makes use of all the exploration 
    rate parameters that were defined above in the hyperparameter section.'''
    exploration_rate = min_exploration_rate + (max_exploration_rate - min_exploration_rate) * np.exp(-exploration_decay_rate*episode)
    
    # Add current episode reward to total rewards list and move on to the next episode
    rewards_all_episodes.append(rewards_current_episode)


C:\Users\rubai\anaconda3\lib\site-packages\gym\utils\passive_env_checker.py:233: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
  if not isinstance(terminated, (bool, np.bool8)):


### All episodes training completed
After all episodes are finished you now just calculate the average reward per thousand episodes from your list that contains the rewards for all episodes so that you can print it out and see how the rewards changed over time.

In [12]:
# Calculate and print the average reward per thousand episodes
rewards_per_thousand_episodes = np.split(np.array(rewards_all_episodes),num_episodes/1000)
count = 1000

print("********Average reward per thousand episodes********\n")
for r in rewards_per_thousand_episodes:
    print(count, ": ", str(sum(r/1000)))
    count += 1000

********Average reward per thousand episodes********

1000 :  0.03200000000000002
2000 :  0.04900000000000004
3000 :  0.06200000000000005
4000 :  0.12800000000000009
5000 :  0.08300000000000006
6000 :  0.07900000000000006
7000 :  0.047000000000000035
8000 :  0.06400000000000004
9000 :  0.03400000000000002
10000 :  0.03000000000000002


### Interpretation

From this print, you can see that the average reward per thousand episodes did indeed progress over time. When the algorithm first started training, the first thousand episodes only averaged a reward of almost **0.18**, but by the time it got to its last thousand episodes, the reward drastically improved to almost **0.7**.

Let's take a second to understand how you can interpret these results. Your agent played **10000** episodes. At each time step within an episode, the agent received a reward of **1** if it reached the frisbee, otherwise, it received a reward of **0**. If the agent did indeed reach the frisbee, then the episode finished at that time-step.

Hence, that means for each episode, the total reward received by the agent for the entire episode is either **1** or **0**. So, for the first thousand episodes, you can interpret this score as meaning that **18%** of the time the agent received a reward of **1** and won the episode. And by the last thousand episodes from a total of **10000**, the agent was winning almost **70%** of the time.

By analyzing the grid of the game, you can see it is a lot more likely that the agent would fall in a hole or perhaps reach the max time steps than it is to reach the frisbee, so reaching the frisbee **70%** of the time is not too bad, especially since the agent had no explicit instructions to reach the frisbee. It learned that this is the correct thing to do.

* SFFF
* FHFH
* FFFH
* HFFG

At last, print out your updated Q-table to see how that has transitioned from its initial state of all zeros.

In [13]:
# Print updated Q-table
print("\n\n********Q-table********\n")
print(q_table)



********Q-table********

[[0.88416946 0.89196161 0.88775357 0.8961652 ]
 [0.66255651 0.44596795 0.65921939 0.8961652 ]
 [0.78590534 0.79222727 0.76808785 0.8961652 ]
 [0.60169033 0.42575585 0.73239691 0.8961652 ]
 [0.86321502 0.72490746 0.58933283 0.40788474]
 [0.         0.         0.         0.        ]
 [0.28043244 0.33580271 0.71251989 0.09790104]
 [0.         0.         0.         0.        ]
 [0.48074966 0.41084779 0.7505246  0.85980969]
 [0.43287203 0.87449818 0.52260853 0.67666425]
 [0.80791608 0.54512694 0.30763721 0.31736817]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.59540797 0.39680891 0.91463823 0.52652332]
 [0.86807341 0.96373968 0.88749761 0.88493158]
 [0.         0.         0.         0.        ]]


# Hyperparameter Tuning

## 1. Learning rate=alpha:
- If learning rate=0.2 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
- If learning rate=0.4 and other parameters are kept constant, then average reward for 1000 episodes is 0.315 and for 10,000 episodes is 0.640
-  If learning rate=0.99 and other parameters are kept constant, then average reward for 1000 episodes is 0.14and for 10,000 episodes is 0.36

- If learning rate is increased, then average reward over time is decreasing, which is not optimal and happens for instability in learning

## 2. Discount Rate=gamma:

- If gamma=0.99 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
-  If gamma=0.5 and other parameters are kept constant, then average reward for 1000 episodes is 0.071 and for 10,000 episodes is 0.125
- If learning rate=0.1 and other parameters are kept constant, then average reward for 1000 episodes is 0 and for 10,000 episodes is 0
- If gamma=1 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.714

- If gamma is increased, then average reward over time is also increasing

## 3. Number of episodes:

- If num_episodes = 10,000 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
-  If num_episodes = 1000 and other parameters are kept constant, then average reward for 100 episodes is 0 and for 1,000 episodes is 0.39
- If num_episodes = 100,000 and other parameters are kept constant, then average reward for 10,000 episodes is 0.6305 and for 100,000 episodes is 0.66809

- If num_episodes is increased, then average reward over time is also increasing!

## 4. Maximum steps per episode

- If max_steps_per_episode = 100 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
- If max_steps_per_episode = 1000 and other parameters are kept constant, then average reward for 1000 episodes is 0.291 and for 10,000 episodes is 0.643
- If max_steps_per_episode = 10 and other parameters are kept constant, then average reward for 1000 episodes is 0 and for 10,000 episodes is 0

- If max_steps_per_episode is increased, then average reward over time is slightly decreasing than the past reference state and if it is decresead to 10, it is dramatically decreased to 0!

## 5. Exploration rate and max exploration rate

- If exploration_rate= 1, max_exploration_rate=1 and other parameters are kept constant, then average reward for 1000 episodes is 0.341 and for 10,000 episodes is 0.701
- If exploration_rate= 0.1, max_exploration_rate=0.1 and other parameters are kept constant, then average reward for 1000 episodes is 0 and for 10,000 episodes is 0
- If exploration_rate= 0.5, max_exploration_rate=0.5 and other parameters are kept constant, then average reward for 1000 episodes is 0.424 and for 10,000 episodes is 0.670
- If exploration_rate= 0.7, max_exploration_rate=0.7 and other parameters are kept constant, then average reward for 1000 episodes is 0.001 and for 10,000 episodes is 0

- If exploration_rate= 1, max_exploration_rate=1 are decreased from 1(max), then average reward over time is behaving in  both upwards and downwards. After decreaseing these to 0.1, the average reward is zero, and later after increasing it to 0.5, the values are acceptable and near to the 1st state where it was set to 1. If we set exploration_rate= 0.7, max_exploration_rate=0.7, then average reward is unexpectedly zero for almost all range of episodes 

## 6. Minimum exploration rate

- If min_exploration_rate=0.01 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
- If  min_exploration_rate=0.1 and other parameters are kept constant, then average reward for 1000 episodes is 0.19 and for 10,000 episodes is 0.081
- If  min_exploration_rate=0.001 and other parameters are kept constant, then average reward for 1000 episodes is 0 and for 10,000 episodes is 0

- If min_exploration_rate=0.01 is increased, then average reward over time is decreasing. After decreaseing from 0.01 to 0.001, the average reward is zero.

## 7. Exploration decay rate

- If exploration_decay_rate = 0.005 and other parameters are kept constant, then average reward for 1000 episodes is 0.338 and for 10,000 episodes is 0.654
- If  exploration_decay_rate =0.001 and other parameters are kept constant, then average reward for 1000 episodes is 0.047 and for 10,000 episodes is 0.016
- If  exploration_decay_rate =0.01 and other parameters are kept constant, then average reward for 1000 episodes is 0 and for 10,000 episodes is 0
- If exploration_decay_rate=0.0005 and other parameters are kept constant, then average reward for 1000 episodes is 0.032 and for 10,000 episodes is 0.030

- If exploration_decay_rate e=0.005 is increased, then average reward over time is decreasing towards zero. After increasing from 0.005 to 0.001, the average reward is decreasing than before.
- If exploration_decay_rate =0.005 is decreased to 0.0005, then average reward over time is avaialble, but not much higher.